File: src/main/java/ru/nucodelabs/gem/app/io/SonetImportManagerImpl.java
package ru.nucodelabs.gem.app.io;

import ru.nucodelabs.data.ves.ExperimentalMeasurement;
import ru.nucodelabs.data.ves.ModelLayer;
import ru.nucodelabs.data.ves.Picket;
import ru.nucodelabs.files.sonet.*;

import java.io.File;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Stream;

class SonetImportManagerImpl implements SonetImportManager {

    @Override
    public Picket loadNameAndExperimentalDataFromEXPFile(File expFile, Picket target) throws Exception {
        String picketName;
        String fileName = expFile.getName();
        if (fileName.endsWith(".EXP") || fileName.endsWith(".exp")) {
            picketName = fileName.substring(0, fileName.length() - 4);
        } else {
            picketName = fileName;
        }

        return Picket.create(
                picketName,
                loadExperimentalDataFromEXPFile(expFile, target).getExperimentalData(),
                target.getModelData()
        );
    }

    private STTFile getSTTFile(File expFile) throws Exception {
        EXPFile expFile1 = new EXPFileParser(expFile).parse();

        Path expFilePath = expFile.toPath();

        return new STTFileParser(new File(
                expFilePath.getParent().toString()
                        + File.separator
                        + expFile1.getSTTFileName()))
                .parse();
    }

    @Override
    public Picket loadExperimentalDataFromEXPFile(File expFile, Picket target) throws Exception {
        EXPFile expFile1 = new EXPFileParser(expFile).parse();
        STTFile sttFile = getSTTFile(expFile);

        List<Integer> sizes =
                Stream.of(
                        sttFile.getAB_2(),
                        sttFile.getMN_2(),
                        expFile1.getAmperage(),
                        expFile1.getVoltage(),
                        expFile1.getResistanceApparent(),
                        expFile1.getErrorResistanceApparent()
                ).map(List::size).toList();

        int minSize = Collections.min(sizes);

        List<ExperimentalMeasurement> expData = new ArrayList<>();
        for (int i = 0; i < minSize; i++) {
            expData.add(ExperimentalMeasurement.create(
                    sttFile.getAB_2().get(i),
                    sttFile.getMN_2().get(i),
                    expFile1.getResistanceApparent().get(i),
                    expFile1.getErrorResistanceApparent().get(i),
                    expFile1.getAmperage().get(i),
                    expFile1.getVoltage().get(i)
            ));
        }

        return Picket.create(target.getName(), expData, target.getModelData());
    }

    @Override
    public Picket loadModelDataFromMODFile(File modFile, Picket target) throws Exception {
        MODFile modFile1 = new MODFileParser(modFile).parse();
        List<Integer> sizes = Stream.of(
                modFile1.getPower(),
                modFile1.getResistance()
        ).map(List::size).toList();

        int minSize = Collections.min(sizes);

        List<ModelLayer> modelData = new ArrayList<>();
        for (int i = 0; i < minSize; i++) {
            modelData.add(ModelLayer.create(
                    modFile1.getPower().get(i),
                    modFile1.getResistance().get(i)));
        }

        return Picket.create(target.getName(), target.getExperimentalData(), modelData);
    }
}

File: src/main/java/ru/nucodelabs/gem/app/io/SonetImportManagerImpl.kt
package ru.nucodelabs.gem.app.io

import ru.nucodelabs.files.sonet.*
import ru.nucodelabs.geo.ves.ExperimentalData
import ru.nucodelabs.geo.ves.ModelLayer
import ru.nucodelabs.geo.ves.Picket
import java.io.File

internal class SonetImportManagerImpl : SonetImportManager {

    private fun parseExpWithStt(expFile: File): Pair<EXPFile, STTFile> {
        val expParsed = EXPFileParser(expFile).parse()
        val expFilePath = expFile.toPath()
        return expParsed to STTFileParser(
            File("${expFilePath.parent}${File.separator}${expParsed.sttFileName}")
        ).parse()
    }

    private fun parseExperimentalDataAndComment(expFile: File): Pair<List<ExperimentalData>, String> {
        val (expParsed, sttParsed) = parseExpWithStt(expFile)

        val minSize = listOf(
            sttParsed.aB_2,
            sttParsed.mN_2,
            expParsed.amperage,
            expParsed.voltage,
            expParsed.resistivityApparent,
            expParsed.errorResistivityApparent
        ).minOf { it.size }

        val expData: MutableList<ExperimentalData> = mutableListOf()
        for (i in 0 until minSize) {
            expData += ExperimentalData(
                ab2 = sttParsed.aB_2[i],
                mn2 = sttParsed.mN_2[i],
                resistivityApparent = expParsed.resistivityApparent[i],
                errorResistivityApparent = expParsed.errorResistivityApparent[i],
                amperage = expParsed.amperage[i],
                voltage = expParsed.voltage[i]
            )
        }
        return expData to passportAsString(expParsed)
    }

    private fun passportAsString(expFile: EXPFile): String {
        return """
            Номер установки: ${expFile.number}
            Дата: ${expFile.date}
            Погода: ${expFile.weather}
            Оператор: ${expFile.operator}
            Интерпретатор: ${expFile.interpreter}
            Проверил ${expFile.checked}
        """.trimIndent()
    }

    override fun fromEXPFile(expFile: File): Picket {
        val fileName = expFile.name

        val picketName = if (fileName.endsWith(".exp", ignoreCase = true)) {
            fileName.substring(0, fileName.length - ".exp".length)
        } else {
            fileName
        }

        val (expData, passport) = parseExperimentalDataAndComment(expFile)
        return Picket(
            name = picketName,
            experimentalData = expData,
            comment = passport
        )
    }

    override fun fromMODFile(modFile: File): List<ModelLayer> {
        val modParsed = MODFileParser(modFile).parse()

        val minSize = listOf(
            modParsed.power,
            modParsed.resistivity
        ).minOf { it.size }

        val modelData: MutableList<ModelLayer> = mutableListOf()
        for (i in 0 until minSize) {
            modelData += ModelLayer(
                power = modParsed.power[i],
                resistivity = modParsed.resistivity[i]
            )
        }
        return modelData
    }
}